internal/batching/batching.go

```diff
@@ -0,0 +1,272 @@
+func (b Batches) Recycle() {
+	for batchesK, batch := range b {
+		for batchK := range batch {
+			delete(batch, batchK)
+		}
+		putPool(batch)
+		delete(b, batchesK)
+	}
+	putPool(b)
+}
+
+// Iter returns a channel that iterates over the data. Closing ctx will stop the iteration.
+func (b Batches) Iter(ctx context.Context) <-chan data.Entry {
+	ch := make(chan data.Entry, 1)
+	go func() {
+		defer close(ch)
+		for _, batch := range b {
+			for _, d := range batch {
+				select {
+				case <-ctx.Done():
+					return
+				case ch <- d:
+				}
+			}
+		}
+	}()
+	return ch
+}
+
+// Batch is a map of UIDs to data.
+type Batch map[types.UID]data.Entry
+
+func (b *Batch) Map() map[types.UID]data.Entry {
+	if b == nil {
+		return nil
+	}
+	return *b
+}
+
+// Batcher is used to ingest data and emit batches.
+type Batcher struct {
+	timespan  time.Duration
+	current   Batches
+	batchSize int
+
+	in  <-chan data.Entry
+	out chan Batches
+
+	emitter func()
+
+	log *slog.Logger
+}
+
+// Option is a opional argument for New().
+type Option func(*Batcher) error
+
+// WithLogger sets the logger.
+func WithLogger(log *slog.Logger) Option {
+	return func(b *Batcher) error {
+		b.log = log
+		return nil
+	}
+}
+
+// WithBatchSize sets the batch size at which to emit at. So if you set this to 1000, it will
+// emit when it has 1000 items in the batch if we haven't hit the timespan. If the timespan
+// is hit, it will emit regardless of the batch size.
+func WithBatchSize(size int) Option {
+	return func(b *Batcher) error {
+		if size < 0 {
+			return errors.New("batch size must be greater than 0")
+		}
+		b.batchSize = size
+		return nil
+	}
+}
+
+// New creates a new Batcher.
+func New(ctx context.Context, in <-chan data.Entry, out chan Batches, timespan time.Duration, options ...Option) (*Batcher, error) {
+	if in == nil || out == nil {
+		return nil, errors.New("can't call Batcher.New() with a nil in or out channel")
+	}
+
+	b := &Batcher{
+		timespan: timespan,
+		in:       in,
+		out:      out,
+		log:      slog.Default(),
+	}
+	b.current = getBatches()
+	b.emitter = b.emit
+
+	for _, o := range options {
+		if err := o(b); err != nil {
+			return nil, err
+		}
+	}
+
+	go b.run()
+
+	return b, nil
+}
+
+// run runs the Batcher loop.
+func (b *Batcher) run() {
+	defer close(b.out)
+
+	timer := time.NewTimer(b.timespan)
+	ticker := time.NewTicker(b.timespan)
+	defer ticker.Stop()
+
+	for {
+		timer.Reset(b.timespan)
+		exit, err := b.handleInput(timer.C)
+		if err != nil {
+			b.log.Error(err.Error())
+		}
+		if exit {
+			return
+		}
+	}
+}
+
+// handleInput handles the input data and batching when the ticker fires.
+func (b *Batcher) handleInput(tick <-chan time.Time) (exit bool, err error) {
+	select {
+	case data, ok := <-b.in:
+		if !ok {
+			return true, nil
+		}
+		if err := b.handleData(data); err != nil {
+			return false, err
+		}
+		if b.batchSize > 0 && len(b.current) == b.batchSize {
```

Comment:
Instead of just item count, I wonder if we also want to keep track of an estimated payload size?

Reply:
I suppose we'll have an estimate of pod, node, and namespace size after we remove managed fields, though

Reply:
So you can't estimate what the size of an object is until it is marshalled within Go itself.  There is an unsafe way of looking at object sizes, but it only counts pointer size and not what pointers point at.

So it will only matter once we get down the ARG.  And it looks to me that any level of batching is going to be far greater than thresholds by grid, so we are always going to be shipping to storage.  Certainly in large clusters. 

If we have to batch out by size, the ARG connector will be where that happens.

Reply:
Yep, I know there's already some checks there because of needing to send to storage.